### PR TITLE
✂️ [DPO] Fix truncation `keep_end` leading to zero'd out samples

### PR DESCRIPTION
**Pull request by LeonEricsson**
# What does this PR do?
Fixes issue #3382. To recap, performing a left flush followed by left truncation (when `truncation_mode='keep_end'`) will  remove real tokens before masked ones. In the worst case, an entire sample may be masked out, leading to NaNs and potentially corrupting the loss calculation (unconfirmed).

https://github.com/huggingface/trl/blob/999acd53ec8913b3ad49e5b3528e40281659dcb3/trl/trainer/dpo_trainer.py#L1118-L1134

To address this we now:

- ~~**Flush-left** each sequence as before, then~~
- ~~**For keep_end**, compute each row’s actual flushed length _Lᵢ_ and gather the slice `[max(0, Lᵢ–max_length) ... Lᵢ)` individually-so even very short completions still include their real tokens;~~
- ~~**For keep_start**, simply take `[:, :max_length]` on the already left-aligned data.~~

Implement `flush_right` and apply the flush direction depending on the truncation mode (while making sure the final outcome is left flushed)

This per‐row windowing ensures every example contributes at least its own completion tokens, preventing any fully zeroed `loss_mask` rows.

Also, while tinkering I ended up with an optimized version of `flush_left()`, should be **significantly** faster.

### Benchmarks
Some runs on trl-lib/ultrafeedback_binarized with 

```
max_length=512, 
max_prompt_length=256,
truncation_mode='keep_end'
``` 

shows that we no longer have any masked out samples, and that our token utilization has more than doubled (note that the rate of improvement depends heavily on the above parameters and the dataset). 

Screenshot 2025-05-03 at 16 47 57

*The ratio of trainable tokens is calculated by dividing the total number of tokens by the total number of unmasked tokens in the `loss_mask`.*

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
That's a good point @LeonEricsson!
Another (maybe more intuitive?) to think of it:

```python
if self.truncation_mode == "keep_end": 
    attention_mask, input_ids, loss_mask = flush_left(attention_mask, input_ids, loss_mask)
    input_ids = input_ids[:, -self.max_length :]
    attention_mask = attention_mask[:, -self.max_length :]
    loss_mask = loss_mask[:, -self.max_length :]
elif self.truncation_mode == "keep_start":
    attention_mask, input_ids, loss_mask = flush_right(attention_mask, input_ids, loss_mask)
    input_ids = input_ids[:, : self.max_length]
    attention_mask = attention_mask[:, : self.max_length]
    loss_mask = loss_mask[:, : self.max_length]
    attention_mask, input_ids, loss_mask = flush_left(attention_mask, input_ids, loss_mask)
```

Is this roughly equivalent to what you do?

**Comment by qgallouedec** [comment]
Yes I think the latest is probably clearer

**Comment by LeonEricsson** [comment]
> Yes I think the latest is probably clearer

Done

**Review by qgallouedec — APPROVED** [review]
Thanks @LeonEricsson, super clean PR!
